### PR TITLE
create celestia proving service

### PR DESCRIPTION
**Pull request by cmwaters**
## Overview

This rust program runs a gRPC service exposing a small API for generating state transition proofs from tendermint headers as well as membership proofs from IAVL proofs

### CODE REVIEW

**Comment by cmwaters** [comment]
I'm still missing the membership proofs (I've just done the tendermint header proofs), but I will do that in a follow up PR

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,15 @@
...
+# Added by cargo
+
+/target
+
+
+# Added by cargo
+#
+# already existing elements were commented out
+
+#/target
```
Comment (rootulp) [inline .gitignore]:
```suggestion
```

File: ibc/lightclients/groth16/client_state.go [hunk 2]
```diff
@@ -44,7 +45,7 @@ func (cs ClientState) ClientType() string {
 }
 
 // GetLatestHeight returns latest block height.
```
Comment (rootulp) [inline ibc/lightclients/groth16/client_state.go:47]:
```suggestion
// GetLatestClientHeight returns latest block height.
```

File: proto/ibc/lightclients/groth16/v1/groth16.proto [hunk 3]
```diff
@@ -2,23 +2,20 @@ syntax = "proto3";
 
 package celestia.ibc.lightclients.groth16.v1;
 
-option go_package = "github.com/celestiaorg/celestia-zkevm-ibc-demo/ibc/lightclients/groth16";
+option go_package = "ibc/lightclients/groth16";
```
Comment (rootulp) [inline proto/ibc/lightclients/groth16/v1/groth16.proto:5]: note to self, this file was moved which may have fixed https://github.com/celestiaorg/celestia-zkevm-ibc-demo/issues/20 although `make run-demo` no longer exists. And `make proto-gen` fails on this PR so unclear

File: src/main.rs [hunk 4]
```diff
@@ -0,0 +1,3 @@
+fn main() {
+    println!("Hello, world!");
+}
```
Comment (rootulp) [inline src/main.rs:3]: should we delete this file b/c seems like boilerplate that is no longer needed
```suggestion
```

File: Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+reqwest = { version = "0.12", features = ["json"] }
+alloy = { version = "0.8.1", features = ["providers"] }
+alloy-primitives = "0.8"
+alloy-provider = { version = "0.8.1", features = ["default"] }
+tendermint-rpc = "0.40"
+ibc-client-tendermint-types = "0.56"
+
+ibc-eureka-solidity-types = { git = "https://github.com/cosmos/solidity-ibc-eureka.git", branch = "main", subdir = "packages/solidity", features = ["rpc"]}
+sp1-ics07-tendermint-prover = { git = "https://github.com/cosmos/solidity-ibc-eureka.git", branch = "main", subdir = "packages/prover"}
+sp1-ics07-tendermint-utils = { git = "https://github.com/cosmos/solidity-ibc-eureka.git", branch = "main", subdir = "packages/utils"}
```
Comment (rootulp) [inline Cargo.toml:28]: [question] when I run `cargo run` locally, I observe these warnings:

```
$ cargo run
warning: /Users/rootulp/git/rootulp/celestiaorg/celestia-zkevm-ibc-demo/Cargo.toml: unused manifest key: workspace.dependencies.ibc-eureka-solidity-types.subdir
warning: /Users/rootulp/git/rootulp/celestiaorg/celestia-zkevm-ibc-demo/Cargo.toml: unused manifest key: workspace.dependencies.sp1-ics07-tendermint-prover.subdir
warning: /Users/rootulp/git/rootulp/celestiaorg/celestia-zkevm-ibc-demo/Cargo.toml: unused manifest key: workspace.dependencies.sp1-ics07-tendermint-utils.subdir
```

so maybe they are removable (?)

**Review by rootulp — APPROVED** [review]
Should we include a default `.env` in `provers/celestia-prover` with a populated RPC_URL that is the default one used when running `make start`? 

Asking b/c I hit a panic:

```
thread 'main' panicked at provers/celestia-prover/src/main.rs:38:43:
RPC_URL not set: NotPresent
```

File: Cargo.toml (hunk 5 above)
Comment (cmwaters) [reply Cargo.toml:28]: we definitely depend on them so not sure about that

File: proto/ibc/lightclients/groth16/v1/groth16.proto (hunk 3 above)
Comment (cmwaters) [reply proto/ibc/lightclients/groth16/v1/groth16.proto:5]: Good point, let me update the docs. Generation should work by just running buf generate

File: README.md [hunk 6]
```diff
@@ -74,3 +74,15 @@ docker logs reth
...
+
+## Contributing
+
+### Proto Generation
+
+This repo uses protobuf to define the interfaces between several services. To help with this, this
+repo relies on [buf](https://buf.build). If you modify the protos you can regenerate them using:
+
+```
+buf generate
```
Comment (rootulp) [inline README.md:87]: [question] Given this PR also updates the `make proto-gen` command, is it safe to use that here instead?

```suggestion
make proto-gen
```

File: provers/celestia-prover/README.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
+# Celestia Prover
+
+The Celestia Prover is a gRPC service that generates zero-knowledge proofs for Celestia state transitions and data membership. It is designed to work with IBC (Inter-Blockchain Communication) and specifically implements proofs compatible with the ICS-07 Tendermint client specification.
+
+## Usage
+
+> ⚠️ **Warning**
+> This gRPC service is still under development and may not work as described
```
Comment (rootulp) [inline provers/celestia-prover/README.md:8]: [optional] GitHub now supports a special syntax to make this warning pretty in GitHub UI. 

```suggestion
> [!WARNING] 
> This gRPC service is still under development and may not work as described
```

Ref: https://github.com/orgs/community/discussions/16925

File: .env.example [hunk 8]
```diff
@@ -0,0 +1,15 @@
+# SP1_PROVER={network|local|mock}
```
Comment (rootulp) [inline .env.example:1]: [no change needed] should we create a follow-up to de-duplicate this with the .env file that is created at `./solidity-ibc-eureka/.env` via the `setup-env.go`

File: provers/celestia-prover/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    membership_prover: SP1ICS07TendermintProver<MembershipProgram>,
+    evm_rpc_url: Url,
+    evm_contract_address: Address,
+}
+
+impl ProverService {
+    fn new() -> ProverService {
+        let rpc_url = env::var("RPC_URL").expect("RPC_URL not set");
+        let contract_address = env::var("CONTRACT_ADDRESS").expect("CONTRACT_ADDRESS not set");
+        let url = Url::parse(rpc_url.as_str()).expect("Failed to parse URL");
```
Comment (rootulp) [inline provers/celestia-prover/src/main.rs:41]: uber optional

```suggestion
        let url = Url::parse(rpc_url.as_str()).expect("Failed to parse RPC_URL");
```

File: provers/celestia-prover/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+        println!("proving from height {:?} to height {:?}", &trusted_light_block.signed_header.header.height, &proposed_header.trusted_height);
+
+        let proof = self.tendermint_prover.generate_proof(
+            &client_state,
+            &trusted_consensus_state,
+            &proposed_header,
+            now,
+        );
+        
+        // Implement your state transition proof logic here
```
Comment (ninabarbakadze) [inline provers/celestia-prover/src/main.rs:95]: what logic would we implement here?

**Review by rootulp — APPROVED** [review]

**Review by ninabarbakadze — APPROVED** [review]
lgtm. still wrapping my head around exactly how the proving works in the sp1 tendermint package

File: provers/celestia-prover/src/main.rs (hunk 10 above)
Comment (cmwaters) [reply provers/celestia-prover/src/main.rs:95]: Ah I think that can be removed

File: provers/celestia-prover/README.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+To run the server you will need to clone the repo and install rust and cargo. To run the node you also need to set the following environment variables:
+
+- `TENDERMINT_RPC_URL` - the url of the tendermint chain you are proving.
+- `RPC_URL` the json rpc url of the evm chain you are generating the proofs for.
+- `CONTRACT_ADDRESS` - the evm address of the tendermint sp1 ics07 contract.
+
+To then run the server (on port `:50001`):
```
Comment (ninabarbakadze) [inline provers/celestia-prover/README.md:16]: This is wrong

```suggestion
To then run the server (on port `:50051`):
```

Comment (cmwaters) [reply provers/celestia-prover/README.md:16]: good catch

File: provers/celestia-prover/src/main.rs [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+use alloy::providers::ProviderBuilder;
+use alloy::primitives::Address;
+use ibc_core_commitment_types::merkle::MerkleProof;
+
+ pub struct ProverService {
+    tendermint_prover: SP1ICS07TendermintProver<UpdateClientProgram>,
+    tendermint_rpc_client: HttpClient,
+    membership_prover: SP1ICS07TendermintProver<MembershipProgram>,
+    evm_rpc_url: Url,
+    evm_contract_address: Address,
```
[the remark this answers is not in the document]
Comment (cmwaters) [reply provers/celestia-prover/src/main.rs:34]: as in ics07_tendermint_contract_address?

**Pull request merged**